### PR TITLE
Adds new blog post Code-Dive-Openstack-Live-Migrations

### PR DESCRIPTION
**Pull request by Spazmotic**
In a comment on the pull request, include the following information and check off the list.

 - [x] Mini bio
  - Soft capped to 75 words
 - [x] Desired date of publishing
  - Make sure this matches the `date` in the front matter
 - [ ] Social media contact info
 - [ ] Redaction: (yes|no)
  - Allow redaction, editing or structure changes by technical writer staff
 - [x] Comments: (yes|no)
  - If you're willing to answer comments about your post, set `comments: true`.
  - Make sure to subscribe to your post comments by clicking on the **Subscribe** link at the very bottom of the post _once_ it's published.

@catlook Please review.

### CODE REVIEW

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-7259a77d5d/).

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-7259a77d5d/).

**Comment by rackernexus** [comment]
Your content preview [is now ready](https://staging.developer.rackspace.com/build-274585ba01/). :bow:

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-548b5c64f4/).

**Comment by rackernexus** [comment]
:mag_right: [Preview](https://staging.developer.rackspace.com/build-317cb4da80/)

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/build-317cb4da80/)

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-073b68608c/).

**Comment by rackernexus** [comment]
Your content preview [is now ready](https://staging.developer.rackspace.com/build-100641225e/). :bow:

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-804831154d/).

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/build-9ca9087708/)

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/build-1d4f772a3b/)

**Comment by rackernexus** [comment]
Your content preview [is now ready](https://staging.developer.rackspace.com/build-1d4f772a3b/). :bow:

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/build-440cacc290/)

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-440cacc290/).

**Comment by rackernexus** [comment]
:mag_right: [Preview](https://staging.developer.rackspace.com/build-27fd5932c2/)

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-773ee2a067/).

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/build-30fe08d499/)

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-f94a502d69/).

**Comment by rackernexus** [comment]
:mag_right: [Preview](https://staging.developer.rackspace.com/build-0928a61002/)

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/build-0928a61002/)

**Comment by rackernexus** [comment]
:mag_right: [Preview](https://staging.developer.rackspace.com/build-6b866cdf08/)

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/build-9c7b735857/)

File: _posts/2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3.md [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+
+### Exploration 1
+
+![Compute Exploration Flow]({% asset_path 2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3/Compute-1.png %})
+
+To begin we start as always, by transferring the RPC call into a private method.  Some additional work here is handled here however.  Because this process can take a significant amount of time to run, we do not want RPC to be hung up waiting for the \_do\_live\_migration method to return, so we spawn an Eventlet thread which will continue to run, and RPC will return its worker to the pool.  From here, \_do\_live\_migration is essentially told to _fork_ off and handle its business, and we continue into that method.
```
Comment (catlook) [inline _posts/2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3.md]: ...  method. However, Some additional work here is handled here. ...

...  for the *\_do\_live\_migration* method  ...   [ed note: put asterixes around method names (can keep slashes)]

... return. Therefore,  we spawn an Eventlet thread which continues to run, and RPC returns its worker ...

... here, *\_do\_live\_migration* is essentially ...

File: _posts/2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3.md [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+            with self._live_migration_semaphore:
+                self._do_live_migration(*args, **kwargs)
+        ...
+{% endhighlight %}
+
+#### Exploration 2
+
+![Compute Exploration Flow]({% asset_path 2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3/Compute-2.png %})
+
+This is the beginning of the prep phase for the live migration proper. The migration status is set to `Preparing`, and because I'm running a block\_migration, We gather details about the disk information and get basic BDM information again (remember, We discarded this information during the scheduling phase and only returned host details). We pull that very basic information and throw it into the *pre\_live\_migration* through *computeRPC* once again. I save the result of this as *migrate\_data*.  Note here that I actually pass in the current *migrate\_data*, which was once the *dest\_check\_data*, so I'm overwriting that variable with the result of *pre\_live\_migration* method return. Let's look briefly into the *get\_disk\_info* method here for clarity. Once the *pre\_live\_migration* method finishes, we return to this method to continue.
```
Comment (catlook) [inline _posts/2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3.md]: ...  a *block\_migration*, I gather  ...   [ed note: add asterixes around block_migration]
... (remember, I discarded  ...
...  details). I pull ...

File: _posts/2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3.md [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+                    #*#*#*# destination, volumes attached, and various other information has been pulled,
+                    #*#*#*# it would have been a monumental waste of time to have mapping failure at this point.
+                    
+            post_method(context, instance, destination_hostname, 
+                        block_migration, migrate_data)
+ {% endhighlight %}                        
+                        
+The way that *post\_method* is called is a bit interesting, because it is defined by the incoming variables accepted by the *self.driver.live\_migration* method. You can see above that this is called with *self.post\_live\_migration* and *self.\_rollback\_live\_migration*. When a fault occurs, it calls that *recover\_method*, and, in this case, it calls the *post\_method*, which resolves to the *nova.compute.manager.\_post\_live\_migration* method. It is tempting to believe that this is the *nova.virt.xenapi.vmops.post\_live\_migration* method but remember that this data was sent from COMPUTE and not VIRT.  You can also see that the *post\_live\_migration* in XenAPI is *post\_live\_migration* and not *\_post\_live\_migration*. So, let's take a look!
```
Comment (catlook) [inline _posts/2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3.md]: ... the *nova.virt.xenapi.vmops.post\_live\_migration* method, but remember that ....

File: _posts/2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3.md [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+        self.network_api.setup_networks_on_host(context, instance, self.host)
+ {% endhighlight %}
+
+#### Exploration 12
+
+![Compute Exploration Flow]({% asset_path 2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3/Compute-12.png %})
+
+Now we can return to *\_post\_live\_migration* again, with more trimmed down psuedocode in the following sample to see where we left off marked with an (X).
+    
+The prevoius process essentially completed the migration. As far as the database is concerned, the `task_state` is empty, so this thing is no longer running. The instance points to the new host in the database, and *instance.host* = *self.host* (set from the destination). The remainder of the code is handling the last of the cleanup and ending the process completely. Comments are inline.
```
Comment (catlook) [inline _posts/2018-04-20-Code-Dive-Openstack-Live-Migrations-Part-3.md]: The previous process ...

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/build-19120e9dcd/)

**Comment by rackernexus** [comment]
HULK [PREVIEW](https://staging.developer.rackspace.com/build-19120e9dcd/)

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/build-19120e9dcd/).

**Comment by rackernexus** [comment]
:mag_right: [Preview](https://staging.developer.rackspace.com/build-19120e9dcd/)

**Comment by rackernexus** [comment]
Your content preview [is now ready](https://staging.developer.rackspace.com/build-5ef13425fe/). :bow:

**Comment by rackernexus** [comment]
Your content preview [is now ready](https://staging.developer.rackspace.com/build-5ef13425fe/). :bow:

**Pull request merged**